core/src/main/java/hudson/model/AbstractItem.java

```diff
@@ -351,6 +353,11 @@ protected void checkRename(@Nonnull String newName) throws Failure {
      * you can use this method.
      */
     protected void renameTo(final String newName) throws IOException {
+
+        if (!isNameEditable()) {
+            throw new IOException("Trying to rename an item that has not editable name.");
```

Comment:
Should probably be throwing `Failure`, like `doConfirmRename` does. Or better yet, just move the call to `isNameEditable` into `doCheckNewName` where all the other checks are.

Reply:
I did that at the beginning, then I removed it because of @oleg-nenashev comment on it. I prefer to keep a protection here, because this is were action are really done, and we don't know how code will evolve in future or how class that inherit from AbstractItem will use it. I feel it's safer to check just before the real renaming.

For the form `Failure` in `doConfirmRename`, I can redo it again, I also think that it's better for user experience.

Reply:
Likely there is no UX here, since no GUI option to rename would have been offered to begin with if the item were not renamable.

No strong opinion about the placement or type of exception; an error is an error. I do not think anything external actually calls this method directly. Just check for CLI or REST usages.